lib/formatter.ex

```diff
@@ -0,0 +1,289 @@
+  output will contain one object per line.  No trailing newline is emitted.
+
+  The `:record_separator` option may be given to control the string
+  used as newline (default `"\n"`).  Other options are ignored.
+
+  Example:
+
+      iex> Jason.Formatter.minimize(~s|{ "a" : "b" , "c": \n\n 2}|)
+      ~s|{"a":"b","c":2}|
+  """
+  @spec minimize(iodata, opts) :: binary
+  def minimize(iodata, opts \\ []) do
+    minimize_to_iolist(iodata, opts)
+    |> :erlang.list_to_binary
+  end
+
+
+  @doc ~S"""
+  Returns an iolist containing a minimized representation of
+  JSON-encoded `iodata`.
+
+  See `minimize/2` for details and options.
+  """
+  @spec minimize_to_iolist(iodata, opts) :: iolist
+  def minimize_to_iolist(iodata, opts) do
+    pretty_print_to_iolist(
+      iodata,
+      indent: "",
+      line_separator: "",
+      record_separator: opts[:record_separator] || "\n",
+      after_colon: ""
+    )
+  end
+
+
+  ## Returns a copy of `opts` with defaults applied
+  @spec normalize_opts(keyword) :: opts
+  defp normalize_opts(opts) do
+    [
+      indent: opts[:indent] || "  ",
+      line_separator: opts[:line_separator] || "\n",
+      record_separator: opts[:record_separator] || opts[:line_separator] || "\n",
+      after_colon: opts[:after_colon] || " ",
+    ]
+  end
+
+
+  ## Returns an iolist containing `depth` instances of `opts[:indent]`
+  @spec tab(opts, non_neg_integer, iolist) :: iolist
+  defp tab(opts, depth, output \\ []) do
+    if depth < 1 do
+      output
+    else
+      tab(opts, depth-1, [opts[:indent] | output])
+    end
+  end
+
+
+  @typep pp_state :: {
+    non_neg_integer,  ## depth -- current nesting depth
+    boolean,          ## in_str -- is the current byte in a string?
+    boolean,          ## in_bs -- does the current byte follow a backslash in a string?
+    boolean,          ## empty -- is the current object or array empty?
+    boolean,          ## first -- is this the first object or array in the input?
+  }
+
+  @spec pp_iodata(
+    iodata,           ## input -- input data
+    iolist,           ## output_acc -- output iolist (built in reverse order)
+    non_neg_integer,  ## depth -- current nesting depth
+    boolean,          ## in_str -- is the current byte in a string?
+    boolean,          ## in_bs -- does the current byte follow a backslash in a string?
+    boolean,          ## empty -- is the current object or array empty?
+    boolean,          ## first -- is this the first object or array in the input?
+    opts
+  ) :: {iolist, pp_state}
+  defp pp_iodata(input, output_acc, depth, in_str, in_bs, empty, first, opts)
+
+  defp pp_iodata("", output_acc, depth, in_str, in_bs, empty, first, opts) do
+    {:lists.reverse(output_acc), {depth, in_str, in_bs, empty, first, opts}}
+  end
+
+  defp pp_iodata([], output_acc, depth, in_str, in_bs, empty, first, opts) do
+    {:lists.reverse(output_acc), {depth, in_str, in_bs, empty, first, opts}}
```

Comment:
When using iodata, there should generally be no need to reverse at the end, since you can build the result already reversed. You can append to iodata using `[acc | new_data]`.

Reply:
Appending with `|` doesn't work in the general case:

    [100] |> IO.iodata_to_binary  #=> "d"
    [100 | "d"] |> IO.iodata_to_binary  #=> "dd"
    [100 | 100] |> IO.iodata_to_binary  #=> ArgumentError

But, benchmarking `[acc, new_data]` vs. the old `[new_data | acc]`-then-reverse behavior, there's a benefit to building the iodata forwards (mostly above 500kB or so, there's up to a 50% speed increase), so I've changed that.